docs/elasticsearch/ux.md

```diff
@@ -0,0 +1,1081 @@
+
+#### Sample Request
+
+```json
+GET shakespeare/_search
+{
+  "query": {
+    "match_phrase_prefix": {
+      "text_entry": {
+        "query": "qui",
+        "slop": 3
+      }
+    }
+  }
+}
+```
+
+Prefix matching doesn’t require any special mappings. It works with your data as-is.
+However, it’s a fairly resource-intensive operation. A prefix of `a` could match hundreds of thousands of terms and not be useful to your user.
+To limit the impact of prefix expansion, set `max_expansions` to a reasonable number. To learn about the `max_expansions` option, see [Options](../full-text/#options).
+
+#### Sample Request
+
+```json
+GET shakespeare/_search
+{
+  "query": {
+    "match_phrase_prefix": {
+      "text_entry": {
+        "query": "qui",
+        "slop": 3,
+        "max_expansions": 10
+      }
+    }
+  }
+}
+```
+
+The ease of implementing query-time autocomplete comes at the cost of performance.
+When implementing this feature on a large scale, we recommend an index-time solution. With an index-time solution, you might experience slower indexing, but it’s a price you pay only once and not for every query.
+
+### Edge N-gram matching
+
+During indexing, edge N-grams chop up a word into a sequence of N characters to support a faster lookup of partial search terms.
+
+If you N-gram the word "quick," the results depend on the value of N.
+
+N | Type | N-gram
+:--- | :--- | :---
+1 | Unigram | [ `q`, `u`, `i`, `c`, `k` ]
+2 | Bigram | [ `qu`, `ui`, `ic`, `ck` ]
+3 | Trigram | [ `qui`, `uic`, `ick` ]
+4 | Four-gram | [ `quic`, `uick` ]
+5 | Five-gram | [ `quick` ]
+
+Autocomplete needs only the beginning N-grams of a search phrase, so Elasticsearch uses a special type of N-gram called edge N-gram.
+
+Edge N-gramming the word "quick" results in the following:
+
+- `q`
+- `qu`
+- `qui`
+- `quic`
+- `quick`
+
+This follows the same sequence the user types.
+
+To configure a field to use edge N-grams, create an autocomplete analyzer with an `edge_ngram` filter:
+
+#### Sample Request
+
+```json
+PUT shakespeare
+{
+  "mappings": {
+    "properties": {
+      "text_entry": {
+        "type": "text",
+        "analyzer": "autocomplete"
+      }
+    }
+  },
+  "settings": {
+    "analysis": {
+      "filter": {
+        "edge_ngram_filter": {
+          "type": "edge_ngram",
+          "min_gram": 1,
+          "max_gram": 20
+        }
+      },
+      "analyzer": {
+        "autocomplete": {
+          "type": "custom",
+          "tokenizer": "standard",
+          "filter": [
+            "lowercase",
+            "edge_ngram_filter"
+          ]
+        }
+      }
+    }
+  }
+}
+```
+
+This example creates the index and instantiates the edge N-gram filter and analyzer.
+
+The `edge_ngram_filter` produces edge N-grams with a minimum N-gram length of 1 (a single letter) and a maximum length of 20. So, it can offer suggestions for words of up to 20 letters.
```

Comment:
So it offers suggestions for words of up to 20 letters.

Reply:
Done